index.js

```diff
@@ -124,6 +151,298 @@
+      switch (arg.type) {
+        case 'UNARY':  return [arg.$1];
+        case 'BINARY': return [arg.$1, arg.$2];
+        default:       return [];
+      }
+    };
+    
+    //  insert :: [Token] -> Token -> Token -> Integer -> Void
+    var insert = function(children, target, newChild, offset) {
+      for (var idx = 0; idx < children.length; idx += 1) {
+        if (children[idx] === target) {
+          children.splice(idx + offset, 0, newChild);
+          break;
+        }
+      }
+    };
+    
+    //  insertAfter :: [Token] -> Token -> Token -> Void
+    var insertAfter = function(children, target, newChild) {
+      insert(children, target, newChild, 1);
+    };
+    
+    //  insertBefore :: [Token] -> Token -> Token -> Void
+    var insertBefore = function(children, target, newChild) {
+      insert(children, target, newChild, 0);
+    };
+    
+    //  addPrefix :: a -> Token -> Token -> Void
+    var addPrefix = function(acc, token, parent) {
+      var prefix = rule(parent, 'prefix');
+      if (!isEmpty(prefix) && !isDelimiter(token)) {
+        insertBefore(parent.tokens, token, Token('DELIMITER', prefix, []));
+      }
+    };
+    
+    //  addSeparator :: a -> Token -> Token -> Void
+    var addSeparator = function(acc, token, parent) {
+      if (!isLeaf(token)) {
+        var separator = rule(token, 'separator');
+        if (separator !== '') {
+          token.tokens.push(Token('DELIMITER', separator, []));
+        }
+      }
+    };
+    
+    //  addWrapper :: a -> Token -> Token -> Void
+    var addWrapper = function(acc, token, parent) {
+      var open = rule(parent, 'open');
+      if (!isEmpty(open) && isContainer(token) && isContainer(parent)) {
+        insertBefore(parent.tokens, token, Token('DELIMITER', open, []));
+      }
+      var close = rule(parent, 'close');
+      if (!isEmpty(close) && isContainer(token) && isContainer(parent)) {
+        insertAfter(parent.tokens, token, Token('DELIMITER', close, []));
+      }
+    };
+    
+    //  addGlue :: a -> Token -> Token -> Token -> Void
+    var addGlue = function(acc, t0, t1, parent) {
+      var glue = rule(parent, 'glue');
+      if (!isEmpty(glue) && !any([t0, t1], isDelimiter)) {
+        insertAfter(parent.tokens, t0, Token('DELIMITER', glue, []));
+      }
+    };
+    
+    //  setStart :: Integer -> Token -> Token -> Integer
+    var setStart = function(acc, token, parent) {
+      token.start = acc;
+      return acc + token.text.length;
+    };
+    
+    //  setEnd :: Integer -> Token -> Token -> Integer
+    var setEnd = function(acc, token, parent) {
+      token.end = isLeaf(token) ? token.start + token.text.length : acc;
+      return token.end;
+    };
+    
+    // create signature from user `def` params
+    return function(name, constraints, expTypes, expRetType) {
+      
+      var sig = Token('SIG', '', [
+        Name(name),
+        Constraints(constraints),
+        Args(expTypes),
+        Ret(expRetType),
+      ]);
```

Comment:
First we create tokens based on user input. At this point if we printed out our sig it would look like `fromMaybeaMaybeaa`.